compiler/noirc_evaluator/src/ssa/opt/brillig_entry_points.rs

```diff
@@ -80,40 +80,92 @@ impl Ssa {
-                let new_inner_call =
-                    new_functions_map.get(&inner_call).copied().unwrap_or(inner_call);
+            let function = self
+                .functions
+                .get_mut(&new_function_to_update)
+                .expect("ICE: Function does not exist");
+            let new_function_value_id = function.dfg.import_function(new_id);
+            function.dfg[instruction] =
+                Instruction::Call { func: new_function_value_id, arguments };
+        }
 
+        // Second pass to rewrite the calls sites in the cloned functions
+        // The list of structs mapping call site updates in `calls_to_update` only includes the original function IDs
+        // so we risk potentially not rewriting the call sites within the cloned functions themselves.
+        // The cloned functions we are using are stored within the `new_functions_map`.
+        for (_, new_functions_per_entry) in new_functions_map {
+            for new_function in new_functions_per_entry.values() {
                 let function =
-                    self.functions.get_mut(&new_inner_call).expect("ICE: Function does not exist");
-                update_function_calls(
-                    function,
-                    entry_point,
-                    &mut new_functions_map,
-                    &calls_to_update,
-                );
+                    self.functions.get_mut(new_function).expect("ICE: Function does not exist");
+                resolve_cloned_function_call_sites(function, &new_functions_per_entry);
             }
         }
 
         self
     }
 }
 
+/// Given that we have already rewritten all the call sites among the original SSA,
+/// this function provides a helper for resolving the call sites within cloned functions.
+/// This function will update  a cloned function according to the supplied function mapping.
+/// The function assumes that the supplied mapping is per entry point and handled
+/// by the caller of this method.
+fn resolve_cloned_function_call_sites(
+    // Function that was cloned earlier in the pass to specialize functions
+    // in the original SSA
+    function: &mut Function,
+    // Per entry point, maps (old function -> new function)
+    new_functions_map: &HashMap<FunctionId, FunctionId>,
+) {
+    for block_id in function.reachable_blocks() {
+        #[allow(clippy::unnecessary_to_owned)] // clippy is wrong here
+        for instruction_id in function.dfg[block_id].instructions().to_vec() {
+            let instruction = function.dfg[instruction_id].clone();
+            let Instruction::Call { func: func_value_id, arguments } = instruction else {
+                continue;
+            };
+            let func_value = &function.dfg[func_value_id];
+            let Value::Function(func_id) = func_value else { continue };
+
+            let new_func_id = new_functions_map.get(func_id).copied().unwrap_or(*func_id);
+            if new_func_id == *func_id {
```

Comment:
This `unwrap_or(*func_id)` followed by `if .. ==` seems a bit roundabout: are there expected to be cases where the function exists in the `new_functions_map` and the `new_func_id == func_id`?

Otherwise, could use the same pattern as above, i.e.:
```rust
let Some(new_func_id) = new_functions_map.get(func_id).copied() else { continue };
```

Reply:
Switched to using the `let ... else` statement